### ISSUE TITLE
kube-router not cleaning up iptables

### ISSUE THREAD

**Issue by coldnfire** [cloudnativelabs/kube-router#1131]
**What happened?**
This is a follow up to #1021

20 days after upgrade the 1.2.3v to 1.3.0v DNS resolution seems to be impossible through a pod inside the DNS cluster and outside the www.

By the way after each restart of a pod, any service exposed behind an ingress was impossible to access.

I would like to precise that DNS resolution is working from the host and after many restart of coredns or kube-router-* anything changes.

**What did you expect to happen?**

**How can we reproduce the behavior you experienced?**
1. Step 1

`kubectl apply -f https://raw.githubusercontent.com/cloudnativelabs/kube-router/master/daemonset/kubeadm-kuberouter.yaml`

2. Step 2

I try to run this command after i see the k3s log:

`journalctl -u k3s.services -f :`

`Jul 23 17:17:24 velen k3s[1004]: E0723 17:17:24.342599    1004 network_policy_controller.go:229] Aborting sync. Failed to cleanup stale iptables rules: Failed to flush the rules in chain KUBE-NWPLCY-DEFAULT due to running [/usr/sbin/iptables -t filter -X KUBE-NWPLCY-DEFAULT --wait]: exit status 1: iptables: Too many links.`

`velen ~ > kubectl exec -n kube-system -ti kube-router-5s4z8 /sbin/iptables -- /usr/local/bin/kube-router --cleanup-config`

```
Defaulted container "kube-router" out of: kube-router, install-cni (init)
I0723 17:47:56.376036   22157 network_policy_controller.go:525] Cleaning up iptables configuration permanently done by kube-router
E0723 17:47:56.392537   22157 network_policy_controller.go:586] Failed to cleanup iptables rules: running [/sbin/iptables -t filter -X KUBE-POD-FW-OMQL4K23QJXQT6MN --wait]: exit status 1: iptables: Too many links.
I0723 17:47:56.392633   22157 network_services_controller.go:2329] Cleaning up IPVS configuration permanently
E0723 17:47:56.422364   22157 network_services_controller.go:638] Failed to run iptables command: running [/sbin/iptables -t filter -D INPUT -m comment --comment handle traffic to IPVS service IPs in custom chain -m set --match-set kube-router-service-ips dst -j KUBE-ROUTER-SERVICES --wait]: exit status 2: iptables v1.8.7 (legacy): Set kube-router-service-ips doesn't exist.
```

```
Try `iptables -h' or 'iptables --help' for more information.
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x1754051]
```

```
goroutine 1 [running]:
sync.(*Mutex).Lock(...)
	/usr/local/go/src/sync/mutex.go:74
github.com/cloudnativelabs/kube-router/pkg/controllers/proxy.(*NetworkServicesController).cleanupIpvsFirewall(0xc00053fae8)
	/build/pkg/controllers/proxy/network_services_controller.go:654 +0x1d1
github.com/cloudnativelabs/kube-router/pkg/controllers/proxy.(*NetworkServicesController).Cleanup(0xc00053fae8)
	/build/pkg/controllers/proxy/network_services_controller.go:2353 +0x2fc
github.com/cloudnativelabs/kube-router/pkg/cmd.CleanupConfigAndExit()
	/build/pkg/cmd/kube-router.go:65 +0xa5
main.Main(0xc00006c748, 0x31e6920)
	/build/cmd/kube-router/kube-router.go:63 +0x54e
main.main()
	/build/cmd/kube-router/kube-router.go:19 +0x26
command terminated with exit code 2
```

**Screenshots / Architecture Diagrams / Network Topologies **
If applicable, add those here to help explain your problem.

** System Information (please complete the following information):**
 - Kube-Router Version: Running kube-router version v1.3.0, built on 2021-07-05T21:05:58+0000, go1.16.5
 - Kube-Router Parameters:
```
         args:
        - --run-router=true
        - --run-firewall=true
        - --run-service-proxy=false
        - --bgp-graceful-restart=true
```
 - Kubernetes Version: 
 ```
Client Version: version.Info{Major:"1", Minor:"21", GitVersion:"v1.21.3", GitCommit:"ca643a4d1f7bfe34773c74f79527be4afd95bf39", GitTreeState:"clean", BuildDate:"2021-07-15T21:04:39Z", GoVersion:"go1.16.6", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"21", GitVersion:"v1.21.2+k3s1", GitCommit:"5a67e8dc473f8945e8e181f6f0b0dbbc387f6fca", GitTreeState:"clean", BuildDate:"2021-06-21T20:52:44Z", GoVersion:"go1.16.4", Compiler:"gc", Platform:"linux/amd64"}
```

 - Cloud Type: on premise
 - Kubernetes Deployment Type: k3s
 - Kube-Router Deployment Type: DaemonSet
 - Cluster Size: 4
 
** Logs, other output, metrics **
```
Jul 23 17:56:24 velen k3s[1004]: E0723 17:56:24.595812    1004 upgradeaware.go:387] Error proxying data from client to backend: read tcp 10.7.0.57:6443->10.7.0.57:57626: read: connection reset by peer
Jul 23 17:57:57 velen k3s[1004]: E0723 17:57:57.366762    1004 upgradeaware.go:401] Error proxying data from backend to client: write tcp 10.7.0.57:6443->10.7.0.57:57726: write: broken pipe
Jul 23 17:41:45 velen k3s[1004]: E0723 17:41:45.965936    1004 network_policy_controller.go:229] Aborting sync. Failed to cleanup stale iptables rules: Failed to flush the rules in chain KUBE-NWPLCY-DEFAULT due to running [/usr/sbin/iptables -t filter -X KUBE-NWPLCY-DEFAULT --wait]: exit status 1: iptables: Too many links.
```

**Additional context**

Take care )

**Comment by aauren** [comment]
Thanks for reporting. `--cleanup-config` isn't as thorough as it should be for sure. However, I think for that issue I'm going to consolidate this ticket with #1021 since it is really just an add-on to that existing issue with a slightly new permutation of the error. However, this makes it easy for the project to track, instead of having a bunch of tickets that are all slightly related.

In terms of the stack track that happens when it attempts to use a nil mutex, that was already reported in #1126 and was fixed in #1129. Since `--cleanup-config` isn't part of the core runtime, we haven't made a bug fix release for it yet, but we should soon.

**Comment by coldnfire** [comment]
Thanks for your answer, take care )
